Durability.md

```diff
@@ -0,0 +1,218 @@
+the safekeepers, to reduce the load on the primary.
+
+
+Magic STONITH fairy
+-------------------
+
+Now that we have a system that works as long as only one primary node is running at a time, how
+do we ensure that?
+
+1. Use etcd to grant a lease on a key. The primary node is only allowed to operate as primary
+   when it's holding a valid lease. If the primary node dies, the lease expires after a timeout
+   period, and a new node is allowed to become the primary.
+
+2. Use S3 to store the lease. S3's consistency guarantees are more lenient, so in theory you
+   cannot do this safely. In practice, it would probably be OK if you make the lease times and
+   timeouts long enough. This has the advantage that we don't need to introduce a new
+   component to the architecture.
+
+3. Use Raft or Paxos, with the WAL safekeepers acting as the Acceptors to form the quorum. The
+   next chapter describes this option.
+
+
+Built-in Paxos
+--------------
+
+The WAL safekeepers act as PAXOS Acceptors, and the Processing nodes
+as both Proposers and Learners.
+
+Each WAL safekeeper holds an Epoch value in addition to the VCL and
+the WAL. Each request by the primary to safekeep WAL is accompanied by
+an Epoch value. If a safekeeper receives a request with Epoch that
+doesn't match its current Accepted Epoch, it must ignore (NACK) it.
+(In different Paxos papers, Epochs are called "terms" or "round
+numbers")
+
+When a node wants to become the primary, it generates a new Epoch
+value that is higher than any previously observed Epoch value, and
+globally unique.
+
+
+Accepted Epoch: 555                VCL                   LSN
+                                     |                     |
+                                     V                     V
+.................ccccccccccccccccccccXXXXXXXXXXXXXXXXXXXXXXX
+Archived WAL       Completed WAL          In-flight WAL
+
+
+Primary node startup:
+
+1. Contact all WAL safekeepers that you can reach (if you cannot
+   connect to a quorum of them, you can give up immediately). Find the
+   latest Epoch among them.
+
+2. Generate a new globally unique Epoch, greater than the latest Epoch
+   found in previous step.
+
+2. Send the new Epoch in a Prepare message to a quorum of
+   safekeepers. (PAXOS Prepare message)
+
+3. Each safekeeper responds with a Promise. If a safekeeper has
+   already made a promise with a higher Epoch, it doesn't respond (or
+   responds with a NACK). After making a promise, the safekeeper stops
+   responding to any write requests with earlier Epoch.
+
+4. Once you have received a majority of promises, you know that the
+   VCL cannot advance on the old Epoch anymore. This effectively kills
+   any old primary server.
+
+5. Find the highest written LSN among the quorum of safekeepers (these
+   can be included in the Promise messages already). This is the new
+   VCL.  If a new node starts the election process after this point,
```

Comment:
Nope. Only after first record of new Epoch will be commited. Read Raft paper carefully.

Reply:
This is based on Paxos, which is slightly different from Raft on that point, if I've understood them correctly. Can you give a step-by-step sequence of events of where that would happen?

Reply:
For simplicity lets have 3 safekeepers A, B, C
Primary P1 writes records for Epoch E1: R1_1, R1_2 (in format of `R<EPOCH>_<LSN>`)
R1_1 replicated to all A, B, C
R1_2 replicated to B
Primary had got a pause
New primary P2 has promises for epoch E2 from A and B. It designated VCL as R1_2. It copies R1_2 to A and replicates R2_3 to A. Then it failed.
New primary P3 has promises for epoch E3 from A and B. It designates VCL as R2_3. And it failed immediately.
New primary P4 has promises for epoch E4 from B and C. It designates VCL as R1_2.

Reply:
Therefore, VCL have to be defined not as "highest written LSN", but as highest "common LSN". And it could advance only after step 6, when WAL is copied from safekeeper with the highest LSN to quorum.

Reply:
Gotcha, thanks! Right, you must finish step 6 before you can consider the "highest LSN" as the durable VCL.